File: packages/ts-host/src/relay/extract.ts
import type { Message, Task } from '@a2a-js/sdk'
import type { A2uiComponent, A2uiSnapshot } from '../types'

/**
 * Чистые хелперы разбора ответа удалённого A2A-агента (Message | Task). Без ALS/NestJS/LangChain —
 * переносимы в любой relay. Подняты из chat-backend `remote-agent-registry`.
 */

type TextPart = { kind: string; text?: string }

function partsText(parts: ReadonlyArray<TextPart>): string {
  return parts
    .filter((p) => p.kind === 'text' && typeof p.text === 'string')
    .map((p) => p.text)
    .join('')
}

/**
 * Текст ответа из `Task`. Авторитет — `status.message` терминального снапшота; text-артефакты
 * СУММИРОВАТЬ с ним нельзя.
 *
 * Почему: A2A штатно разрешает стримить ответ дельтами (`artifact-update` + `append: true`), и
 * text-артефакт в этом случае — ЖИВАЯ ПРОЕКЦИЯ того же самого ответа, а не вторая его часть. Такой
 * агент отдаёт один и тот же текст дважды: дельтами в артефакт и снапшотом в `status.message`.
 * Прежняя склейка обоих каналов печатала пользователю ОТВЕТ ДВАЖДЫ. Баг был латентным: у агентов на
 * `createAgentHost` артефакты несут только `kind:'data'` (см. build-task `toTask`), поэтому склейка
 * была безвредна — и ломалась ровно на том, кто пользуется штатным стримингом A2A.
 *
 * Почему авторитет именно `status.message`: прошлые `artifact-update` сервер НЕ реплеит (журнала
 * событий нет), поэтому после reconnect/`tasks/get` доживает только снапшот — стрим невосстановим.
 *
 * Fallback на артефакты — если терминального текста нет вовсе (агент отдал только стрим): тогда
 * артефакты и есть единственный источник.
 */
function collectTaskText(task: Task): string {
  const statusText = task.status.message?.parts
    ? partsText(task.status.message.parts as TextPart[])
    : ''
  if (statusText) return statusText

  const chunks: string[] = []
  for (const artifact of task.artifacts ?? []) chunks.push(partsText(artifact.parts as TextPart[]))
  return chunks.filter(Boolean).join('\n\n')
}

/** Текст из результата `sendMessage` (Message | Task). */
export function extractText(result: Message | Task): string {
  const text = result.kind === 'task' ? collectTaskText(result) : partsText(result.parts as TextPart[])
  return text.trim()
}

/**
 * A2UI из ответа сабагента: `completed` → `artifact.parts[data].data.a2ui`,
 * `input-required` (форма) → `task.metadata.a2ui`. Элементы — сырые деревья
 * `{component, props, children?, catalogId?}` и/или конверты `A2uiSnapshot`
 * (стабильные id + dataModel, сквозной контракт lookup) — пробрасываются как
 * есть: оркестратор кладёт их в свой `result.a2ui`, host эмитит с теми же id.
 */
export function extractA2ui(result: Message | Task): (A2uiComponent | A2uiSnapshot)[] {
  if (result.kind !== 'task') return []
  const out: (A2uiComponent | A2uiSnapshot)[] = []
  for (const artifact of result.artifacts ?? []) {
    for (const part of artifact.parts) {
      if (part.kind === 'data') {
        const a2ui = (part.data as { a2ui?: unknown } | undefined)?.a2ui
        if (Array.isArray(a2ui)) out.push(...(a2ui as (A2uiComponent | A2uiSnapshot)[]))
      }
    }
  }
  const metaA2ui = (result.metadata as { a2ui?: unknown } | undefined)?.a2ui
  if (Array.isArray(metaA2ui)) out.push(...(metaA2ui as (A2uiComponent | A2uiSnapshot)[]))
  return out
}

/**
 * Ошибка «таск устарел/не найден/в терминальном состоянии» — повод повторить БЕЗ `resumeTaskId`
 * (свежий диалог). Покрывает A2A `TaskNotFoundError` (-32001) и текстовые маркеры.
 */
export function isStaleTaskError(err: unknown): boolean {
  const code = (err as { code?: unknown } | undefined)?.code
  if (code === -32001) return true
  const msg = String((err as { message?: unknown } | undefined)?.message ?? err ?? '').toLowerCase()
  return (
    (msg.includes('task') &&
      (msg.includes('not found') || msg.includes('final') || msg.includes('terminal'))) ||
    msg.includes('cannot be continued')
  )
}

File: packages/python-host/src/ai37_agent_host/relay/extract.py
"""Разбор ответа удалённого A2A-агента — порт ``ts-host/src/relay/extract.ts``.

Чистые хелперы над ``Message | Task``: без ALS/LangChain — переносимы в любой relay. В отличие от
JS (дискриминатор ``result.kind``), в protobuf-мире ``a2a-sdk`` типа-тега нет: Task отличаем по
наличию поля ``status``/``artifacts``. На вход — protobuf-объект ИЛИ уже нормализованный dict
(``MessageToDict``) — для тестируемости. A2UI-деревья пробрасываем как СЫРЫЕ dict'ы (оркестратор
поднимет их своим surface'ом).
"""

from __future__ import annotations

from typing import Any

from google.protobuf.json_format import MessageToDict


def _to_dict(result: Any) -> dict[str, Any]:
    if isinstance(result, dict):
        return result
    if hasattr(result, "DESCRIPTOR"):
        return MessageToDict(result, preserving_proto_field_name=False)
    return {}


def _is_task(d: dict[str, Any]) -> bool:
    return "status" in d or "artifacts" in d


def _parts_text(parts: Any) -> str:
    if not isinstance(parts, list):
        return ""
    return "".join(
        p["text"] for p in parts if isinstance(p, dict) and isinstance(p.get("text"), str)
    )


def _collect_task_text(task: dict[str, Any]) -> str:
    """Текст ответа из ``Task``. Авторитет — ``status.message`` терминального снапшота;
    text-артефакты СУММИРОВАТЬ с ним нельзя (порт фикса ts-host ``relay/extract.ts``).

    Почему: A2A штатно разрешает стримить ответ дельтами (``artifact-update`` + ``append``), и
    text-артефакт в этом случае — ЖИВАЯ ПРОЕКЦИЯ того же ответа, а не вторая его часть. Такой
    агент отдаёт один и тот же текст дважды: дельтами в артефакт и снапшотом в ``status.message``.
    Прежняя склейка обоих каналов печатала пользователю ОТВЕТ ДВАЖДЫ. Баг был латентным: у агентов
    на ``create_agent_host`` артефакты несут только data-part (см. ``a2a_executor._finalize``),
    поэтому склейка была безвредна — и ломалась ровно на том, кто пользуется штатным стримингом A2A.

    Почему авторитет именно ``status.message``: прошлые ``artifact-update`` сервер НЕ реплеит
    (журнала событий нет), поэтому после reconnect/``tasks/get`` доживает только снапшот — стрим
    невосстановим.

    Fallback на артефакты — если терминального текста нет вовсе (агент отдал только стрим): тогда
    артефакты и есть единственный источник.
    """
    status = task.get("status")
    if isinstance(status, dict) and isinstance(status.get("message"), dict):
        status_text = _parts_text(status["message"].get("parts"))
        if status_text:
            return status_text

    chunks: list[str] = []
    for artifact in task.get("artifacts") or []:
        if isinstance(artifact, dict):
            chunks.append(_parts_text(artifact.get("parts")))
    return "\n\n".join(c for c in chunks if c)


def extract_text(result: Any) -> str:
    """Текст из результата ``send_message`` (Message | Task)."""
    d = _to_dict(result)
    text = _collect_task_text(d) if _is_task(d) else _parts_text(d.get("parts"))
    return text.strip()


def extract_a2ui(result: Any) -> list[dict[str, Any]]:
    """Сырые A2UI: ``completed`` → ``artifact.parts[data].data.a2ui``; форма → ``metadata.a2ui``."""
    d = _to_dict(result)
    if not _is_task(d):
        return []
    out: list[dict[str, Any]] = []
    for artifact in d.get("artifacts") or []:
        if not isinstance(artifact, dict):
            continue
        for part in artifact.get("parts") or []:
            if isinstance(part, dict) and isinstance(part.get("data"), dict):
                a2ui = part["data"].get("a2ui")
                if isinstance(a2ui, list):
                    out.extend(a2ui)
    meta = d.get("metadata")
    if isinstance(meta, dict) and isinstance(meta.get("a2ui"), list):
        out.extend(meta["a2ui"])
    return out


def is_stale_task_error(err: Any) -> bool:
    """Таск устарел/не найден/терминален → повторить БЕЗ resume (A2A -32001 + текстовые маркеры)."""
    code = err.get("code") if isinstance(err, dict) else getattr(err, "code", None)
    if code == -32001:
        return True
    raw = err.get("message") if isinstance(err, dict) else getattr(err, "message", None)
    msg = str(raw if raw is not None else err or "").lower()
    return (
        "task" in msg and ("not found" in msg or "final" in msg or "terminal" in msg)
    ) or "cannot be continued" in msg
